pkg/ctrl/map_resource.go

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ctrl
+
+import "k8s.io/client-go/util/workqueue"
+
+var _ EventHandler = MapResource{}
+
+// MapResource maps an object to a collection of keys
+type MapResource struct {
+	// ObjectToObjectKeys transforms the argument into a slice of keys to be reconciled
+	ObjectToKeys ObjectToObjectKeys
+
+	// Predicate implements Predicate, and is used to filter events before they are given to EventHandlers
+	// Defaults to accept all
+	Predicate
```

Comment:
Can the filter be implemented as a separate type of `MapResource` or `ObjectToObjectKeys`?

Reply:
I don't understand.

Reply:
I'm wondering if the predicate has to be coupled (composed) to that layer, or if it can be inserted as a filter layer instead?

Reply:
It could be separate.  Not sure what is more clean or consistent with other libraries.

Reply:
Yeah, no strong opinion. I like to have each component do one thing, but the complexity might not be needed in that case.